core/src/main/java/io/questdb/log/LogAlertSocket.java

```diff
@@ -0,0 +1,325 @@
+        fdSocket = Net.socketTcp(true);
+        System.out.printf("Connecting with: %s:%d%n", hosts[hostPortIdx], ports[hostPortIdx]);
+        if (fdSocket > -1) {
+            if (Net.connect(fdSocket, fdSocketAddress) != 0) {
+                System.out.println(" E could not connect");
+                freeSocket();
+            }
+        } else {
+            System.out.println(" E could not create TCP socket [errno=" + ff.errno() + "]");
+            freeSocket();
+        }
+    }
+
+    public void send(int len) {
+        send(len, 0, null);
+    }
+
+    public void send(int len, Consumer<String> ackReceiver) {
+        send(len, 0, ackReceiver);
+    }
+
+    @Override
+    public void close() {
+        if (fdSocket != -1) {
+            freeSocket();
+        }
+        if (outBufferPtr != 0) {
+            Unsafe.free(outBufferPtr, outBufferSize, MemoryTag.NATIVE_DEFAULT);
+            outBufferPtr = 0;
+        }
+        if (inBufferPtr != 0) {
+            Unsafe.free(inBufferPtr, IN_BUFFER_SIZE, MemoryTag.NATIVE_DEFAULT);
+            inBufferPtr = 0;
+        }
+    }
+
+    public long getOutBufferPtr() {
+        return outBufferPtr;
+    }
+
+    public int getOutBufferSize() {
+        return outBufferSize;
+    }
+
+    public long getInBufferPtr() {
+        return inBufferPtr;
+    }
+
+    public long getInBufferSize() {
+        return inBufferSize;
+    }
+
+    private void send(int len, int failOverLevel, Consumer<String> ackReceiver) {
+        if (fdSocket > 0) {
+            //
+            int remaining = len;
+            long p = outBufferPtr;
+            boolean sendFail = false;
+            while (remaining > 0) {
+                int n = Net.send(fdSocket, p, remaining);
+                if (n > 0) {
+                    remaining -= n;
+                    p += n;
+                } else {
+                    System.out.println("could not send [n=" + n + " [errno=" + ff.errno() + "]");
+                    sendFail = true;
+                }
+            }
+            if (sendFail) {
+                failOver(len, failOverLevel + 1, ackReceiver);
+            } else {
+                // receive ack
+                p = inBufferPtr;
+                final int n = Net.recv(fdSocket, p, inBufferSize);
+                if (n > 0) {
+                    if (ackReceiver != null) {
+                        ackReceiver.accept(Chars.stringFromUtf8Bytes(inBufferPtr, inBufferPtr + n));
+                    } else {
+                        Net.dumpAscii(p, n);
+                    }
+                } else {
+                    failOver(len, failOverLevel + 1, ackReceiver);
+                }
+            }
+        } else {
+            failOver(len, failOverLevel + 1, ackReceiver);
+        }
+    }
+
+    private void failOver(int len, int failOverLevel, Consumer<String> ackReceiver) {
+        freeSocket();
+        hostPortIdx = (hostPortIdx + 1) % hostPortSize;
```

Comment:
when only one target server is specified the mechanism will probably fail with `array index out of bounds`

Reply:
in this case `hostPortSize` will == 1, so failOver will be yet again the same host.

The scenario I had in mind:
- There will be 1..N alert managers forming a cluster
- their host:port list is known to the alert socket writer
- At least 1 of these alert managers is guaranteed to be up and running and available because kubctl ensures this

So the scenario where there is only one configured alert manager is mostly for development. In this case, we trampoline with the recursive function and attempt to run again eagerly, limiting the number of attempts to 100.

I will change the recursive method.